WordPress/Classes/ViewRelated/Post/PostEditorState.swift

```diff
@@ -0,0 +1,292 @@
+        case .draft:
+            return PostEditorStateSave()
+        default:
+            return self
+        }
+    }
+
+    func updated(publishDate: Date?, context: PostEditorStateContext) -> PostEditorActionState {
+        guard let publishDate = publishDate else {
+            return self
+        }
+
+        if isFutureDated(publishDate) {
+            return PostEditorStateSchedule()
+        }
+
+        return self
+    }
+}
+
+fileprivate class PostEditorStateSave: PostEditorActionState {
+    var action: PostEditorAction {
+        return .save
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        switch postStatus {
+        case .publish:
+            return PostEditorStateUpdate()
+        default:
+            return self
+        }
+    }
+
+    func updated(publishDate: Date?, context: PostEditorStateContext) -> PostEditorActionState {
+        guard let publishDate = publishDate else {
+            return self
+        }
+
+        if isFutureDated(publishDate) {
+            return PostEditorStateSchedule()
+        }
+
+        return self
+    }
+}
+
+fileprivate class PostEditorStateSchedule: PostEditorActionState {
+    var action: PostEditorAction {
+        return .schedule
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        return self
+    }
+
+    func updated(publishDate: Date?, context: PostEditorStateContext) -> PostEditorActionState {
+        return self
+    }
+}
+
+fileprivate class PostEditorStateSubmitForReview: PostEditorActionState {
+    var action: PostEditorAction {
+        return .submitForReview
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        return self
+    }
+
+    func updated(publishDate: Date?, context: PostEditorStateContext) -> PostEditorActionState {
+        return self
+    }
+}
+
+fileprivate class PostEditorStateUpdate: PostEditorActionState {
+    var action: PostEditorAction {
+        return .update
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        return self
+    }
+
+    func updated(publishDate: Date?, context: PostEditorStateContext) -> PostEditorActionState {
+        guard let publishDate = publishDate else {
+            return self
+        }
+
+        if isFutureDated(publishDate) && context.originalPostStatus != .scheduled {
+            return PostEditorStateSchedule()
+        }
+
+        if isPastDated(publishDate) && context.originalPostStatus == .scheduled {
+            return PostEditorStatePublish()
+        }
+
+        return self
+    }
+}
+
+fileprivate extension PostEditorActionState {
+    func isFutureDated(_ date: Date) -> Bool {
+        let oneMinute: TimeInterval = 60.0
+        let dateOneMinuteFromNow = Date(timeInterval: oneMinute, since: Date())
```

Comment:
I would instead use `Calendar.current.date(byAdding: .minute, value: 1, to: date)`

Or you could potentially use `Calendar.current.compare(Date(), to: date, toGranularity: .minute)` to compare them instead.